dbms/src/Functions/FunctionsString.cpp

```diff
@@ -5956,6 +5956,256 @@ class FunctionFormatDecimal : public IFunction
+
+    template <typename IntType>
+    static bool executeUnHexInt(
+        const ColumnPtr & column,
+        const bool is_unsigned,
+        ColumnString::Chars_t & res_chars,
+        ColumnString::Offsets & res_offsets,
+        ColumnUInt8::Container & res_null_map)
+    {
+        const auto col = checkAndGetColumn<ColumnVector<IntType>>(column.get());
+        if (col == nullptr)
+        {
+            return false;
+        }
+        const size_t size = col->size();
+        res_chars.resize(size * 10);
+        res_offsets.resize(size);
+
+        char low;
+        char high;
+        char data[20];
+        size_t length;
+        size_t pos = 0;
+        for (size_t i = 0; i < size; ++i)
+        {
+            if (is_unsigned)
+            {
+                UInt64 number = col->getUInt(i);
+                length = sprintf(data, "%lu", number);
+            }
+            else
+            {
+                Int64 number = col->getInt(i);
+                if (number < 0)
+                {
+                    pos++;
+                    res_offsets[i] = pos;
+                    res_null_map[i] = 1;
+                    continue;
+                }
+                length = sprintf(data, "%li", number);
+            }
+
+            size_t begin = 0;
+            if (length % 2 != 0)
+            {
+                const char * byte = reinterpret_cast<const char *>(&data[begin]);
+                fromHexChar(byte, low);
+                res_chars[pos] = low;
+                pos++;
+                begin++;
+            }
+            for (size_t j = begin; j < length; j += 2)
+            {
+                const char * byte1 = reinterpret_cast<const char *>(&data[j]);
+                const char * byte2 = reinterpret_cast<const char *>(&data[j + 1]);
+                fromHexChar(byte1, high);
+                fromHexChar(byte2, low);
+                res_chars[pos] = (high << 4) | low;
+                pos++;
+            }
+            pos++;
+            res_offsets[i] = pos;
+        }
+        res_chars.resize(pos);
+
+        return true;
+    }
+
+    static void unhexOne(const ColumnString::Chars_t & data,
+                         const size_t length,
+                         const size_t idx,
+                         size_t begin,
+                         size_t pos,
+                         ColumnString::Chars_t & res_data,
+                         ColumnString::Offsets & res_offsets,
+                         ColumnUInt8::Container & res_null_map)
+    {
+        char low;
+        char high;
+        size_t end = begin + length;
+        res_offsets[idx] = pos + 1;
+
+        if (length % 2 != 0)
+        {
+            const char * byte = reinterpret_cast<const char *>(&data[begin]);
+            if (!fromHexChar(byte, low))
+            {
+                res_null_map[idx] = 1;
+                return;
+            }
+            res_data[pos] = low;
+            pos++;
+            begin++;
+        }
```

Comment:
Why we need to handle this situation?

Reply:
`Unhex` uses two characters to represent the high four bits and the low four bits. If the length of the string is odd, the first character needs to be processed separately.